README.md
<a href="#installation" width="100%">
<img src="banner.svg" alt="Title banner"/>
</a>

## Tailwind CSS Stripes Plugin

[![npm version](https://badge.fury.io/js/@designbycode%2Ftailwindcss-stripes.svg)](https://badge.fury.io/js/@designbycode%2Ftailwindcss-stripes)
![npm](https://img.shields.io/npm/dt/%40designbycode/tailwindcss-stripes)
![NPM](https://img.shields.io/npm/l/%40designbycode%2Ftailwindcss-stripes)
![npm bundle size](https://img.shields.io/bundlephobia/min/%40designbycode%2Ftailwindcss-stripes)
![ts](https://badgen.net/badge/Built%20With/TypeScript/blue)
[![GitHub stars](https://img.shields.io/github/stars/DesignByCode/tailwindcss-stripes?style=social)](https://github.com/DesignByCode/tailwindcss-stripes/stargazers)
[![HitCount](https://hits.dwyl.com/designbycode/tailwindcss-stripes.svg?style=flat)](http://hits.dwyl.com/designbycode/tailwindcss-stripes)

The Tailwind CSS Stripes Plugin provides a set of utility classes and components that allow you to easily create striped backgrounds for your HTML elements. It offers a flexible way to customize the color, size, opacity, and animation
direction of the stripes.

![Stripes Plugin](screen.png)

## 📇 Table of Contents

* [Installation](#installation)
    * [Using pnpm](#using-pnpm)
    * [Using npm](#using-npm)
    * [Using yarn](#using-yarn)
* [Configuration](#configuration)
* [Usage](#usage)
    * [Utility Classes](#utility-classes)
        * [Stripes Utility](#stripes-utility)
        * [Stripes Color Modifier](#stripes-color-modifier)
        * [Stripes Direction Modifier](#stripes-direction-modifier)
        * [Stripes Opacity Modifier](#stripes-opacity-modifier)
        * [Stripes Size Modifier](#stripes-size-modifier)
* [Customization](#customization)
* [Conclusion](#conclusion)
* [Contributing](#contributing)
* [License](#license)
* [Author](#author)
* [Acknowledgments](#acknowledgments)

## Installation

To use the Tailwind CSS Stripes Plugin, you need to have Tailwind CSS installed in your project. If you don't have it yet, you can follow the installation guide to set it up.

Once you have Tailwind CSS installed, you can add the Stripes Plugin to your project using pnpm, npm or yarn:

#### Using pnpm

```bash
pnpm add @designbycode/tailwindcss-stripes
```

#### Using npm

```bash
npm install @designbycode/tailwindcss-stripes
```

#### Using yarn

```bash
yarn add @designbycode/tailwindcss-stripes
```

## Configuration

To enable the Stripes Plugin, you need to add it to your Tailwind CSS configuration file. Open your tailwind.config.js and import the plugin at the top of the file:

```javascript
module.exports = {
    // Your existing configuration...
    plugins: [
        // Other plugins...
        require("@designbycode/tailwindcss-stripes"),
    ],
};
```

## Usage

Once the Stripes Plugin is installed and added to your configuration, you can start using the utility classes and components to create striped backgrounds.

### Utility Classes

The plugin provides the following utility classes:

#### Stripes Utility

```.stripes```
The ```.stripes``` class creates a container with relative positioning, isolates its contents from being affected by the animation, and hides any overflowing content. It serves as the base class for creating the striped effect.

```html

<div class="stripes">
    <!-- Your content here -->
</div>
```

#### Stripes Color Modifier

```.stripes-{value}```

The ```.stripes-{value}``` class sets the color of the stripes to selected value. You can use it to create any color stripes on your elements.

```html

<div class="stripes stripes-rose-500">
    <!-- Your content here -->
</div>

<!-- by color name or arbitrary value -->

<div class="stripes stripes-[chocolate] stripes-[#ff000]">
    <!-- Your content here -->
</div>
```

#### Stripes Direction Modifier

```.stripes-reverse```

The ```.stripes-reverse``` class reverses the direction of the stripes' animation. It creates a cool effect where the stripes appear to move in the opposite direction.

```html

<div class="stripes stripes-reverse">
    <!-- Your content here -->
</div>
```

#### Stripes Opacity Modifier

```.stripes-opacity-{value}```

The ```.stripes-opacity-{value}``` classes allow you to control the opacity of the stripes. Replace {value} with the desired opacity level. The available values are defined in your Tailwind CSS opacity configuration.

```html

<div class="stripes stripes-opacity-50">
    <!-- Your content here -->
</div>

<!-- or by arbitrary value -->
<div class="stripes stripes-opacity-[0.12]">
    <!-- Your content here -->
</div>
```

#### Stripes Size Modifier

```.stripes-size-{value}```

The ```.stripes-size-{value}``` classes let you adjust the size of the stripes. Replace {value} with one of the custom stripe sizes defined in your Tailwind CSS stripeSizes configuration.

```html

<div class="stripes stripes-size-md">
    <!-- Your content here -->
</div>
<!-- or by arbitrary size value -->
<div class="stripes stripes-size-[42px]">
    <!-- Your content here -->
</div>
```

## Customization

You can customize the plugin by modifying the theme section in your Tailwind CSS configuration file. The plugin allows you to define custom stripe sizes and use them as utility classes.

```javascript
module.exports = {
    theme: {
        extend: {
            stripeSizes: {
                // Define your custom stripe sizes here
                xs: "4px",
                sm: "8px",
                md: "16px",
                lg: "24px",
                xl: "32px",
                "2xl": "40px",
                "3xl": "48px",
            },
        },
    },
    plugins: [
        // Other plugins...
        stripesPlugin,
    ],
};

```

By adding or modifying the custom stripe sizes, you can use the .stripes-size-{value} utility classes with your defined sizes.

## Conclusion

The Tailwind CSS Stripes Plugin simplifies the process of creating striped backgrounds with utility classes and components. You can easily apply striped patterns to your elements using the provided classes and customize their appearance
according to your project's needs. The plugin offers control over colors, sizes, opacity, and animation direction, making it a powerful addition to your Tailwind CSS toolkit.

## Contributing

Contributions to this plugin are welcome! If you encounter any issues, have feature requests, or want to improve the plugin, feel free to create a pull request or submit an issue on the GitHub repository.

## License

This project is licensed under the [MIT](LICENCE) License - see the [LICENSE](LICENCE) file for details.

## Author

<div>
<img  align="left" style="box-shadow:3px 3px 3px rgba(0,0,0,75);border-radius:1rem;border:solid 2px rgba(255,225,225,.25)" src="https://github.com/designbycode.png?size=130" alt="Claude Myburgh">
</div>
<h2 style="margin-top:0">Claude Myburgh</h2><ul style="padding-left:0;margin-top:-.63rem;list-style:none"><li>Github:<a href="https://github.com/designbycode"> @designbycode</a></li><li>Npm:<a href="https://www.npmjs.
com/~designbycode_"> @designbycode_</a></li></ul>

## Acknowledgments

- This plugin is inspired by the needs of web developers using Tailwind CSS.
- Special thanks to the Tailwind CSS team for creating such an amazing framework.







CHANGELOG.md
# Changelog

## [Unreleased]

## [v1.0.0] - 2023-08-06

### Added or Changed

- Added this changelog :)

## [v1.0.1] - 2023-08-06

### Fixed

- Replaced tailwindcss @apply with vanilla css (#6fd75)

### Added

- README file added


index.js
const plugin = require("tailwindcss/plugin");

const { convertToRgb } = require("./utils/convert-to-rgb");

module.exports = plugin(
  function ({
    e,
    addBase,
    addComponents,
    addUtilities,
    matchUtilities,
    theme,
  }) {
    const themeColors = theme("colors");

    addBase({
      "@keyframes slides": {
        from: { transform: "translateX(0)" },
        to: { transform: "translateX(var(--stripes-size))" },
      },
      ":root": {
        "--stripes-rgb": "0 0 0",
        "--stripes-angle": "-45deg",
        "--stripes-speed": "1s",
        "--stripes-direction": "forwards",
        "--stripes-opacity": "1",
        "--stripes-size": "20px",
        "--stripes-width": "10%",
      },
    });

    addComponents({
      ".stripes": {
        position: "relative",
        overflow: "hidden",
        isolation: "isolate",
      },
      ".stripes:before": {
        "--stripes-color": "rgb(var(--stripes-rgb) / var(--stripes-opacity))",
        content: '""',
        position: "absolute",
        top: "0",
        right: "0",
        width: "calc(100% + var(--stripes-size))",
        height: "100%",
        pointerEvents: "none",
        backgroundImage: `linear-gradient(
          var(--stripes-angle),
          var(--stripes-color) 5%,
          transparent 5%,
          transparent 45%,
          var(--stripes-color) 45%,
          var(--stripes-color) 55%,
          transparent 55%,
          transparent 95%,
          var(--stripes-color) 95%
          )`,
        backgroundSize: "var(--stripes-size) var(--stripes-size)",
        animation:
          "var(--stripes-speed) slides infinite linear var(--stripes-direction)",
      },
    });

    addUtilities({
      ".stripes-reverse": { "--stripes-direction": "reverse" },
    });

    matchUtilities(
      {
        "stripes-opacity": (value) => ({
          "--stripes-opacity": value,
        }),
      },
      {
        values: theme("opacity"),
      }
    );

    matchUtilities(
      {
        "stripes-size": (value) => ({
          "--stripes-size": value,
        }),
      },
      {
        values: theme("stripeSizes"),
      }
    );

    addUtilities(
      Object.keys(themeColors).reduce((stripes, key) => {
        if (typeof themeColors[key] === "string") {
          return {
            ...stripes,
            [`.stripes-${e(key)}`]: {
              "--stripes-rgb": convertToRgb(themeColors[key]),
            },
          };
        }
        if (typeof themeColors[key] === "function") {
          return {
            ...stripes,
            [`.stripes-${e(key)}`]: {
              "--stripes-rgb": convertToRgb(themeColors[key]({})),
            },
          };
        }

        const colorShades = Object.keys(themeColors[key]);

        return {
          ...stripes,
          ...colorShades.reduce(
            (a, shade) => ({
              ...a,
              [`.stripes-${e(key)}-${shade}`]: {
                "--stripes-rgb": convertToRgb(themeColors[key][shade]),
              },
            }),
            {}
          ),
        };
      }, {})
    );

    matchUtilities(
      {
        stripes: (value) => {
          return {
            "--stripes-rgb": `${convertToRgb(value)}`,
          };
        },
      },
      {
        values: theme("stripeColors"),
      }
    );
  },
  {
    theme: {
      stripeSizes: {
        xm: "6px",
        sm: "12px",
        md: "20px",
        lg: "30px",
        xl: "40px",
        "2xl": "50px",
        "3xl": "60px",
      },
    },
  }
);


utils/color-names.js
export default {
    aliceblue: [240, 248, 255],
    antiquewhite: [250, 235, 215],
    aqua: [0, 255, 255],
    aquamarine: [127, 255, 212],
    azure: [240, 255, 255],
    beige: [245, 245, 220],
    bisque: [255, 228, 196],
    black: [0, 0, 0],
    blanchedalmond: [255, 235, 205],
    blue: [0, 0, 255],
    blueviolet: [138, 43, 226],
    brown: [165, 42, 42],
    burlywood: [222, 184, 135],
    cadetblue: [95, 158, 160],
    chartreuse: [127, 255, 0],
    chocolate: [210, 105, 30],
    coral: [255, 127, 80],
    cornflowerblue: [100, 149, 237],
    cornsilk: [255, 248, 220],
    crimson: [220, 20, 60],
    cyan: [0, 255, 255],
    darkblue: [0, 0, 139],
    darkcyan: [0, 139, 139],
    darkgoldenrod: [184, 134, 11],
    darkgray: [169, 169, 169],
    darkgreen: [0, 100, 0],
    darkgrey: [169, 169, 169],
    darkkhaki: [189, 183, 107],
    darkmagenta: [139, 0, 139],
    darkolivegreen: [85, 107, 47],
    darkorange: [255, 140, 0],
    darkorchid: [153, 50, 204],
    darkred: [139, 0, 0],
    darksalmon: [233, 150, 122],
    darkseagreen: [143, 188, 143],
    darkslateblue: [72, 61, 139],
    darkslategray: [47, 79, 79],
    darkslategrey: [47, 79, 79],
    darkturquoise: [0, 206, 209],
    darkviolet: [148, 0, 211],
    deeppink: [255, 20, 147],
    deepskyblue: [0, 191, 255],
    dimgray: [105, 105, 105],
    dimgrey: [105, 105, 105],
    dodgerblue: [30, 144, 255],
    firebrick: [178, 34, 34],
    floralwhite: [255, 250, 240],
    forestgreen: [34, 139, 34],
    fuchsia: [255, 0, 255],
    gainsboro: [220, 220, 220],
    ghostwhite: [248, 248, 255],
    gold: [255, 215, 0],
    goldenrod: [218, 165, 32],
    gray: [128, 128, 128],
    green: [0, 128, 0],
    greenyellow: [173, 255, 47],
    grey: [128, 128, 128],
    honeydew: [240, 255, 240],
    hotpink: [255, 105, 180],
    indianred: [205, 92, 92],
    indigo: [75, 0, 130],
    ivory: [255, 255, 240],
    khaki: [240, 230, 140],
    lavender: [230, 230, 250],
    lavenderblush: [255, 240, 245],
    lawngreen: [124, 252, 0],
    lemonchiffon: [255, 250, 205],
    lightblue: [173, 216, 230],
    lightcoral: [240, 128, 128],
    lightcyan: [224, 255, 255],
    lightgoldenrodyellow: [250, 250, 210],
    lightgray: [211, 211, 211],
    lightgreen: [144, 238, 144],
    lightgrey: [211, 211, 211],
    lightpink: [255, 182, 193],
    lightsalmon: [255, 160, 122],
    lightseagreen: [32, 178, 170],
    lightskyblue: [135, 206, 250],
    lightslategray: [119, 136, 153],
    lightslategrey: [119, 136, 153],
    lightsteelblue: [176, 196, 222],
    lightyellow: [255, 255, 224],
    lime: [0, 255, 0],
    limegreen: [50, 205, 50],
    linen: [250, 240, 230],
    magenta: [255, 0, 255],
    maroon: [128, 0, 0],
    mediumaquamarine: [102, 205, 170],
    mediumblue: [0, 0, 205],
    mediumorchid: [186, 85, 211],
    mediumpurple: [147, 112, 219],
    mediumseagreen: [60, 179, 113],
    mediumslateblue: [123, 104, 238],
    mediumspringgreen: [0, 250, 154],
    mediumturquoise: [72, 209, 204],
    mediumvioletred: [199, 21, 133],
    midnightblue: [25, 25, 112],
    mintcream: [245, 255, 250],
    mistyrose: [255, 228, 225],
    moccasin: [255, 228, 181],
    navajowhite: [255, 222, 173],
    navy: [0, 0, 128],
    oldlace: [253, 245, 230],
    olive: [128, 128, 0],
    olivedrab: [107, 142, 35],
    orange: [255, 165, 0],
    orangered: [255, 69, 0],
    orchid: [218, 112, 214],
    palegoldenrod: [238, 232, 170],
    palegreen: [152, 251, 152],
    paleturquoise: [175, 238, 238],
    palevioletred: [219, 112, 147],
    papayawhip: [255, 239, 213],
    peachpuff: [255, 218, 185],
    peru: [205, 133, 63],
    pink: [255, 192, 203],
    plum: [221, 160, 221],
    powderblue: [176, 224, 230],
    purple: [128, 0, 128],
    rebeccapurple: [102, 51, 153],
    red: [255, 0, 0],
    rosybrown: [188, 143, 143],
    royalblue: [65, 105, 225],
    saddlebrown: [139, 69, 19],
    salmon: [250, 128, 114],
    sandybrown: [244, 164, 96],
    seagreen: [46, 139, 87],
    seashell: [255, 245, 238],
    sienna: [160, 82, 45],
    silver: [192, 192, 192],
    skyblue: [135, 206, 235],
    slateblue: [106, 90, 205],
    slategray: [112, 128, 144],
    slategrey: [112, 128, 144],
    snow: [255, 250, 250],
    springgreen: [0, 255, 127],
    steelblue: [70, 130, 180],
    tan: [210, 180, 140],
    teal: [0, 128, 128],
    thistle: [216, 191, 216],
    tomato: [255, 99, 71],
    turquoise: [64, 224, 208],
    violet: [238, 130, 238],
    wheat: [245, 222, 179],
    white: [255, 255, 255],
    whitesmoke: [245, 245, 245],
    yellow: [255, 255, 0],
    yellowgreen: [154, 205, 50],
}


utils/convert-to-rgb.js
import colorNames from "./color-names"

export function convertToRgb(color, theme) {
  try {
    if (typeof color === "object") {
      color = color.DEFAULT
    }

    color = color.replace(/\s/g, "")

    const hexPattern = /^#(?:[0-9a-fA-F]{3}){1,2}(?:[0-9a-fA-F]{2})?$/
    const rgbPattern = /^rgba?\((\d{1,3}),\s?(\d{1,3}),\s?(\d{1,3})(?:,\s?([\d.]+))?\)$/
    const hslPattern = /^hsla?\((\d+),\s?([\d.]+)%,\s?([\d.]+)%,?\s?([\d.]+)?\)$/

    if (colorNames && typeof colorNames === "object") {
      const rgbValue = colorNames[color]
      if (rgbValue) {
        return `${rgbValue[0]} ${rgbValue[1]} ${rgbValue[2]}`
      }
    }

    if (hexPattern.test(color)) {
      let hex = color.slice(1)

      if (hex.length === 3) {
        hex = hex.replace(/./g, (char) => char + char)
      }

      const r = parseInt(hex.slice(0, 2), 16)
      const g = parseInt(hex.slice(2, 4), 16)
      const b = parseInt(hex.slice(4, 6), 16)

      return `${r} ${g} ${b}`
    }

    const rgbMatch = color.match(rgbPattern)
    if (rgbMatch) {
      const r = parseInt(rgbMatch[1], 10)
      const g = parseInt(rgbMatch[2], 10)
      const b = parseInt(rgbMatch[3], 10)

      return `${r} ${g} ${b}`
    }

    const hslMatch = color.match(hslPattern)
    if (hslMatch) {
      const h = parseInt(hslMatch[1], 10)
      const s = parseFloat(hslMatch[2])
      const l = parseFloat(hslMatch[3])

      const c = (1 - Math.abs(2 * l - 1)) * s
      const x = c * (1 - Math.abs(((h / 60) % 2) - 1))
      const m = l - c / 2

      let r, g, b

      if (h >= 0 && h < 60) {
        ;[r, g, b] = [c, x, 0]
      } else if (h >= 60 && h < 120) {
        ;[r, g, b] = [x, c, 0]
      } else if (h >= 120 && h < 180) {
        ;[r, g, b] = [0, c, x]
      } else if (h >= 180 && h < 240) {
        ;[r, g, b] = [0, x, c]
      } else if (h >= 240 && h < 300) {
        ;[r, g, b] = [x, 0, c]
      } else {
        ;[r, g, b] = [c, 0, x]
      }

      r = Math.round((r + m) * 255)
      g = Math.round((g + m) * 255)
      b = Math.round((b + m) * 255)

      return `${r} ${g} ${b}`
    }

    if (theme && typeof theme === "object") {
      const colorValue = theme.colors[color]
      if (colorValue) {
        return convertToRgb(colorValue, theme)
      }
    }

    return null
  } catch (e) {
    throw new Error("Invalid color format. Supported formats: #RRGGBB, #RGB, rgb(255, 0, 128), rgba(255, 0, 128, 0.5), hsl(120, 100%, 50%), hsla(120, 100%, 50%, 0.5)")
  }
}

export function isValidColorFormat(color) {
  const hexPattern = /^#(?:[0-9a-fA-F]{3}){1,2}(?:[0-9a-fA-F]{2})?$/
  const rgbPattern = /^rgba?\((\d{1,3}),\s?(\d{1,3}),\s?(\d{1,3})(?:,\s?([\d.]+))?\)$/
  const hslPattern = /^hsla?\((\d+),\s?([\d.]+)%,\s?([\d.]+)%,?\s?([\d.]+)?\)$/
  return hexPattern.test(color) || rgbPattern.test(color) || hslPattern.test(color)
}


src/index.ts
import { comment } from "postcss"
import { version as packageVersion } from "../package.json"

const plugin = require("tailwindcss/plugin")

const { convertToRgb } = require("../utils/convert-to-rgb")

module.exports = plugin.withOptions(
  function () {
    return function ({ e, addBase, addComponents, addUtilities, matchUtilities, theme }: any) {
      const themeColors = theme("colors")

      addBase([
        comment({
          text: `!  tailwindcss-stripes v${packageVersion} | MIT License | https://designbycode.co.za`,
        }),
      ])

      addBase({
        "@keyframes slides": {
          from: { transform: "translateX(0)" },
          to: { transform: "translateX(var(--stripes-size))" },
        },
        ":root": {
          "--stripes-rgb": "0 0 0",
          "--stripes-angle": "-45deg",
          "--stripes-speed": "1s",
          "--stripes-direction": "forwards",
          "--stripes-opacity": "1",
          "--stripes-size": "20px",
          "--stripes-width": "10%",
        },
      })

      addComponents({
        ".stripes": {
          position: "relative",
          overflow: "hidden",
          isolation: "isolate",
        },
        ".stripes:before": {
          "--stripes-color": "rgb(var(--stripes-rgb) / var(--stripes-opacity))",
          content: '""',
          position: "absolute",
          top: "0",
          right: "0",
          width: "calc(100% + var(--stripes-size))",
          height: "100%",
          pointerEvents: "none",
          backgroundImage: `linear-gradient(var(--stripes-angle), var(--stripes-color) 5%, transparent 5%, transparent 45%, var(--stripes-color) 45%, var(--stripes-color) 55%, transparent 55%, transparent 95%, var(--stripes-color) 95%)`,
          backgroundSize: "var(--stripes-size) var(--stripes-size)",
          animation: "var(--stripes-speed) slides infinite linear var(--stripes-direction)",
        },
      })

      addUtilities({
        ".stripes-reverse": { "--stripes-direction": "reverse" },
      })

      matchUtilities(
        {
          "stripes-opacity": (value: any) => ({
            "--stripes-opacity": value,
          }),
        },
        {
          values: theme("opacity"),
        }
      )

      matchUtilities(
        {
          "stripes-size": (value: any) => ({
            "--stripes-size": value,
          }),
        },
        {
          values: theme("stripeSizes"),
        }
      )

      addUtilities(
        Object.keys(themeColors).reduce((stripes, key) => {
          if (typeof themeColors[key] === "string") {
            return {
              ...stripes,
              [`.stripes-${e(key)}`]: {
                "--stripes-rgb": convertToRgb(themeColors[key]),
              },
            }
          }
          if (typeof themeColors[key] === "function") {
            return {
              ...stripes,
              [`.stripes-${e(key)}`]: {
                "--stripes-rgb": convertToRgb(themeColors[key]({})),
              },
            }
          }

          const colorShades = Object.keys(themeColors[key])

          return {
            ...stripes,
            ...colorShades.reduce(
              (a, shade) => ({
                ...a,
                [`.stripes-${e(key)}-${shade}`]: {
                  "--stripes-rgb": convertToRgb(themeColors[key][shade]),
                },
              }),
              {}
            ),
          }
        }, {})
      )

      matchUtilities(
        {
          stripes: (value: any) => {
            return {
              "--stripes-rgb": `${convertToRgb(value)}`,
            }
          },
        },
        {
          values: theme("stripeColors"),
        }
      )
    }
  },
  function () {
    return {
      theme: {
        stripeSizes: {
          xm: "6px",
          sm: "12px",
          md: "20px",
          lg: "30px",
          xl: "40px",
          "2xl": "50px",
          "3xl": "60px",
        },
      },
    }
  }
)
